### PR TITLE
[BEAM-59] initial interfaces and classes of Beam FileSystem.

### PR DESCRIPTION
**Pull request by peihe**
Be sure to do all of the following to help us incorporate your contribution
quickly and easily:

 - [ ] Make sure the PR title is formatted like:
   `[BEAM-<Jira issue #>] Description of pull request`
 - [ ] Make sure tests pass via `mvn clean verify`. (Even better, enable
       Travis-CI on your fork and ensure the whole test matrix passes).
 - [ ] Replace `<Jira issue #>` in the title with the actual Jira issue
       number, if there is one.
 - [ ] If this contribution is large, please file an Apache
       [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.txt).

---

### CODE REVIEW

**Comment by peihe** [comment]
R: @lukecwik

File: sdks/java/core/src/main/java/org/apache/beam/sdk/io/FileSystems.java [hunk 1]
```diff
@@ -0,0 +1,160 @@
...
+
+  private static final ClassLoader CLASS_LOADER = ReflectHelpers.findClassLoader();
+
+  private static final Map<String, FileSystemRegistrar> REGISTRAR_MAP =
+      Collections.synchronizedMap(new HashMap<String, FileSystemRegistrar>());
+
+  private static final Map<String, PipelineOptions> REGISTRAR_DEFAULT_CONFIG_MAP =
+      Collections.synchronizedMap(new HashMap<String, PipelineOptions>());
+
+  public static void loadFileSystemRegistrars() {
```
Comment (lukecwik) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/io/FileSystems.java]: should this be performed only on class load?

File: sdks/java/core/src/main/java/org/apache/beam/sdk/io/FileSystems.java [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+/**
+ * Clients facing {@link FileSystem} utility.
+ */
+public class FileSystems {
+
+  public static final String DEFAULT_SCHEME = "default";
+
+  private static final Pattern URI_SCHEME_PATTERN = Pattern.compile("^[a-zA-Z][-a-zA-Z0-9+.]*$");
+
+  private static final ClassLoader CLASS_LOADER = ReflectHelpers.findClassLoader();
```
Comment (lukecwik) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/io/FileSystems.java]: If you early bind the classloader, loadFileSystemRegistrars will not use the current threads context class loader.

You should call ReflectHelpers.findClassLoader() each time one is needed.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/io/FileSystems.java [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+            scheme, URI_SCHEME_PATTERN.pattern()));
+    REGISTRAR_DEFAULT_CONFIG_MAP.put(
+        checkNotNull(scheme, "scheme"), checkNotNull(options, "options"));
+  }
+
+  /**
+   * Internal method to get {@link FileSystem} for {@code spec}.
+   */
+  @VisibleForTesting
+  public static FileSystem getFileSystemInternal(URI uri) {
```
Comment (lukecwik) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/io/FileSystems.java]: package private?

File: sdks/java/core/src/main/java/org/apache/beam/sdk/io/FileSystems.java [hunk 4]
```diff
@@ -0,0 +1,160 @@
...
+   * Internal method to get {@link FileSystemRegistrar} for {@code scheme}.
+   */
+  @VisibleForTesting
+  static FileSystemRegistrar getRegistrarInternal(String scheme) {
+    if (REGISTRAR_MAP.containsKey(scheme)) {
+      return REGISTRAR_MAP.get(scheme);
+    } else if (REGISTRAR_MAP.containsKey(DEFAULT_SCHEME)) {
+      return REGISTRAR_MAP.get(DEFAULT_SCHEME);
+    } else {
+      throw new IllegalStateException("Unable to find handler for " + scheme);
```
Comment (lukecwik) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/io/FileSystems.java]: to find handler for -> to find registrar for

File: sdks/java/core/src/main/java/org/apache/beam/sdk/io/FileSystems.java [hunk 5]
```diff
@@ -0,0 +1,160 @@
...
+  public static void loadFileSystemRegistrars() {
+    REGISTRAR_MAP.clear();
+    Set<FileSystemRegistrar> registrars =
+        Sets.newTreeSet(ReflectHelpers.ObjectsClassComparator.INSTANCE);
+    registrars.addAll(Lists.newArrayList(
+        ServiceLoader.load(FileSystemRegistrar.class, CLASS_LOADER)));
+
+    checkDuplicateScheme(registrars);
+
+    for (FileSystemRegistrar registrar : registrars) {
```
Comment (lukecwik) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/io/FileSystems.java]: Build the multimap here instead of deferring to addFileSystemRegistrar, this way all conflicts for a single scheme can be printed out instead of only two of the conflicts.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/io/FileSystems.java [hunk 6]
```diff
@@ -0,0 +1,160 @@
...
+  }
+
+  /**
+   * Adds the {@link FileSystemRegistrar} to the static {@code REGISTRAR_MAP}.
+   *
+   * @throws IllegalStateException if multiple {@link FileSystem FileSystems}
+   * for the same scheme are detected.
+   */
+  @VisibleForTesting
+  private static void addFileSystemRegistrar(FileSystemRegistrar registrar) {
```
Comment (lukecwik) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/io/FileSystems.java]: This method is pointless since checkDuplicateScheme will catch all failures, might as well inline the REGISTRAR_MAP.put() into loadFileSystemRegistrars()

File: sdks/java/core/src/main/java/org/apache/beam/sdk/io/FileSystems.java [hunk 7]
```diff
@@ -0,0 +1,160 @@
...
+              + "Scheme: [%s] is already registered with %s, and override is not allowed.",
+          REGISTRAR_MAP.get(registrar.getScheme()).getClass(),
+          registrar.getScheme(),
+          registrar.getClass()));
+    }
+    REGISTRAR_MAP.put(registrar.getScheme(), registrar);
+  }
+
+  @VisibleForTesting
+  static void checkDuplicateScheme(Set<FileSystemRegistrar> registrars) {
```
Comment (lukecwik) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/io/FileSystems.java]: checkDuplicateScheme -> verifySchemesAreUnique

File: sdks/java/core/src/main/java/org/apache/beam/sdk/io/FileSystems.java [hunk 8]
```diff
@@ -0,0 +1,160 @@
...
+ */
+public class FileSystems {
+
+  public static final String DEFAULT_SCHEME = "default";
+
+  private static final Pattern URI_SCHEME_PATTERN = Pattern.compile("^[a-zA-Z][-a-zA-Z0-9+.]*$");
+
+  private static final ClassLoader CLASS_LOADER = ReflectHelpers.findClassLoader();
+
+  private static final Map<String, FileSystemRegistrar> REGISTRAR_MAP =
```
Comment (lukecwik) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/io/FileSystems.java]: REGISTRAR_MAP -> SCHEME_TO_REGISTRAR

Also, you mention that the map is case insensitive on the scheme yet there is no upper/lower casing of the scheme before insertion or the usage of a case insensitive comparator.

Finally, use a ConcurrentHashMap<> which is specifically built for thread safety instead of Collections.syncMap()

File: sdks/java/core/src/main/java/org/apache/beam/sdk/io/FileSystems.java [hunk 9]
```diff
@@ -0,0 +1,160 @@
...
+      Collections.synchronizedMap(new HashMap<String, FileSystemRegistrar>());
+
+  private static final Map<String, PipelineOptions> REGISTRAR_DEFAULT_CONFIG_MAP =
```
Comment (lukecwik) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/io/FileSystems.java]: REGISTRAR_DEFAULT_CONFIG_MAP -> SCHEME_TO_DEFAULT_CONFIG

Also, you mention that the map is case insensitive on the scheme yet there is no upper/lower casing of the scheme before insertion or the usage of a case insensitive comparator.

Finally, use a ConcurrentHashMap<> which is specifically built for thread safety instead of Collections.syncMap()

File: sdks/java/core/src/main/java/org/apache/beam/sdk/io/FileSystems.java [hunk 10]
```diff
@@ -0,0 +1,160 @@
...
+  }
+
+  /**
+   * Sets the default config for {@link FileSystemRegistrar} that associates with {@code scheme}.
+   *
+   * <p>Syntax: scheme = alpha *( alpha | digit | "+" | "-" | "." )
+   * Upper case letters are treated as the same as lower case letters.
+   */
+  public static void setDefaultConfig(String scheme, PipelineOptions options) {
+    checkArgument(
```
Comment (lukecwik) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/io/FileSystems.java]: Should we also check that there is a registrar for the default config so we avoid simple typos.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/io/FileSystems.java [hunk 11]
```diff
@@ -0,0 +1,160 @@
...
+
+    checkDuplicateScheme(registrars);
+
+    for (FileSystemRegistrar registrar : registrars) {
+      addFileSystemRegistrar(registrar);
+    }
+  }
+
+  /**
+   * Sets the default config for {@link FileSystemRegistrar} that associates with {@code scheme}.
```
Comment (lukecwik) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/io/FileSystems.java]: Sets the default configuration to be used with a {@link FileSystemRegistrar} for the provided {@code scheme}.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/io/FileSystems.java [hunk 12]
```diff
@@ -0,0 +1,160 @@
...
+   *
+   * <p>Syntax: scheme = alpha *( alpha | digit | "+" | "-" | "." )
```
Comment (lukecwik) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/io/FileSystems.java]: Stick the scheme in a pre box or it may get formatted funny

File: sdks/java/core/src/test/java/org/apache/beam/sdk/io/FileSystemsTest.java [hunk 13]
```diff
@@ -0,0 +1,55 @@
...
+import com.google.common.collect.Sets;
+import org.apache.beam.sdk.options.PipelineOptionsFactory;
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.rules.ExpectedException;
+
+/**
+ * Tests for {@link FileSystems}.
+ */
+public class FileSystemsTest {
```
Comment (lukecwik) [inline sdks/java/core/src/test/java/org/apache/beam/sdk/io/FileSystemsTest.java]: missing @RunWith...

File: sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/storage/GcsFileSystemRegistrarTest.java [hunk 14]
```diff
@@ -0,0 +1,57 @@
...
+        : Lists.newArrayList(ServiceLoader.load(FileSystemRegistrar.class).iterator())) {
+      if (registrar instanceof GcsFileSystemRegistrar) {
+        return;
+      }
+    }
+    fail("Expected to find " + GcsFileSystemRegistrar.class);
+  }
+
+  @Test
+  public void testGetFileSystem() throws Exception {
```
Comment (lukecwik) [inline sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/storage/GcsFileSystemRegistrarTest.java]: I don't think this provides much value over a single test in FileSystems for checking that loading works correctly.

Here you just want to validate the scheme is gs and that an instance is created when passed pipeline options

File: sdks/java/core/src/test/java/org/apache/beam/sdk/io/LocalFileSystemRegistrarTest.java [hunk 15]
```diff
@@ -0,0 +1,55 @@
...
+      if (registrar instanceof LocalFileSystemRegistrar) {
+        return;
+      }
+    }
+    fail("Expected to find " + LocalFileSystemRegistrar.class);
+  }
+
+  @Test
+  public void testGetFileSystem() throws Exception {
+    FileSystems.loadFileSystemRegistrars();
```
Comment (lukecwik) [inline sdks/java/core/src/test/java/org/apache/beam/sdk/io/LocalFileSystemRegistrarTest.java]: Here you just want to validate the scheme is file and that an instance is created when passed pipeline options

File: sdks/java/core/src/test/java/org/apache/beam/sdk/io/FileSystemsTest.java [hunk 16]
```diff
@@ -0,0 +1,55 @@
...
+    FileSystems.setDefaultConfig("gs-s3", PipelineOptionsFactory.create());
+  }
+
+  @Test
+  public void testSetDefaultConfigInvalidScheme() throws Exception {
+    thrown.expect(IllegalArgumentException.class);
+    thrown.expectMessage("Scheme: [gs:] doesn't match URI syntax");
+    FileSystems.setDefaultConfig("gs:", PipelineOptionsFactory.create());
+  }
+
```
Comment (lukecwik) [inline sdks/java/core/src/test/java/org/apache/beam/sdk/io/FileSystemsTest.java]: Add the test about special handling logic for local file systems here instead of in localfilesystemstest

File: sdks/java/io/hdfs/src/main/java/org/apache/beam/sdk/io/hdfs/HadoopFileSystem.java [hunk 17]
```diff
@@ -0,0 +1,28 @@
...
+package org.apache.beam.sdk.io.hdfs;
+
+import org.apache.beam.sdk.io.FileSystem;
+
+/**
+ * Implements Beam {@link FileSystem} for {@link org.apache.hadoop.fs.FileSystem} connectors.
+ */
+class HadoopFileSystem implements FileSystem {
+
+  HadoopFileSystem(){}
```
Comment (lukecwik) [inline sdks/java/io/hdfs/src/main/java/org/apache/beam/sdk/io/hdfs/HadoopFileSystem.java]: fix spacing around (){}

File: sdks/java/io/hdfs/src/main/java/org/apache/beam/sdk/io/hdfs/HadoopFileSystem.java [hunk 18]
```diff
@@ -0,0 +1,28 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.beam.sdk.io.hdfs;
+
+import org.apache.beam.sdk.io.FileSystem;
+
+/**
+ * Implements Beam {@link FileSystem} for {@link org.apache.hadoop.fs.FileSystem} connectors.
```
Comment (lukecwik) [inline sdks/java/io/hdfs/src/main/java/org/apache/beam/sdk/io/hdfs/HadoopFileSystem.java]: Adapts {@link org.apache.hadoop.fs.FileSystem} connectors to be used as Apache Beam {@link FileSystem}s.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/io/LocalFileSystem.java [hunk 19]
```diff
@@ -0,0 +1,27 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.beam.sdk.io;
+
+/**
+ * Implements {@link FileSystem} for local files.
```
Comment (lukecwik) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/io/LocalFileSystem.java]: {@link FileSystem} implementation for local files.

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/storage/GcsFileSystem.java [hunk 20]
```diff
@@ -0,0 +1,34 @@
...
+ */
+package org.apache.beam.sdk.io.gcp.storage;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+
+import org.apache.beam.sdk.io.FileSystem;
+import org.apache.beam.sdk.options.GcsOptions;
+
+/**
+ * Implements {@link FileSystem} for Google Cloud Storage.
```
Comment (lukecwik) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/storage/GcsFileSystem.java]: {@link FileSystem} implementation for Google Cloud Storage.

File: sdks/java/io/hdfs/src/main/java/org/apache/beam/sdk/io/hdfs/HadoopFileSystemRegistrar.java [hunk 21]
```diff
@@ -0,0 +1,42 @@
...
+public class HadoopFileSystemRegistrar implements FileSystemRegistrar {
+
+  @Override
+  public FileSystem fromOptions(@Nonnull PipelineOptions options) {
+    return new HadoopFileSystem();
+  }
+
+  @Override
+  public String getScheme() {
+    return FileSystems.DEFAULT_SCHEME;
```
Comment (lukecwik) [inline sdks/java/io/hdfs/src/main/java/org/apache/beam/sdk/io/hdfs/HadoopFileSystemRegistrar.java]: why is this the default scheme instead of hdfs://?

File: sdks/java/io/hdfs/src/test/java/org/apache/beam/sdk/io/hdfs/HadoopFileSystemRegistrarTest.java [hunk 22]
```diff
@@ -0,0 +1,55 @@
...
+      if (registrar instanceof HadoopFileSystemRegistrar) {
+        return;
+      }
+    }
+    fail("Expected to find " + HadoopFileSystemRegistrar.class);
+  }
+
+  @Test
+  public void testGetFileSystem() {
+    FileSystems.loadFileSystemRegistrars();
```
Comment (lukecwik) [inline sdks/java/io/hdfs/src/test/java/org/apache/beam/sdk/io/hdfs/HadoopFileSystemRegistrarTest.java]: Like the other tests, this doesn't provide much value, just create an instance, check the scheme and validate that its configured when passed pipeline options.

File: sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/storage/GcsFileSystemRegistrarTest.java (hunk 14 above)
Comment (peihe) [inline sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/storage/GcsFileSystemRegistrarTest.java]: removed

File: sdks/java/io/hdfs/src/test/java/org/apache/beam/sdk/io/hdfs/HadoopFileSystemRegistrarTest.java (hunk 22 above)
Comment (peihe) [inline sdks/java/io/hdfs/src/test/java/org/apache/beam/sdk/io/hdfs/HadoopFileSystemRegistrarTest.java]: done

File: sdks/java/core/src/test/java/org/apache/beam/sdk/io/LocalFileSystemRegistrarTest.java (hunk 15 above)
Comment (peihe) [inline sdks/java/core/src/test/java/org/apache/beam/sdk/io/LocalFileSystemRegistrarTest.java]: removed

File: sdks/java/core/src/main/java/org/apache/beam/sdk/io/FileSystems.java (hunk 10 above)
Comment (peihe) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/io/FileSystems.java]: done

File: sdks/java/core/src/main/java/org/apache/beam/sdk/io/FileSystems.java (hunk 1 above)
Comment (peihe) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/io/FileSystems.java]: I don't get it.
Only use CLASS_LOADER?

File: sdks/java/core/src/main/java/org/apache/beam/sdk/io/FileSystem.java [hunk 23]
```diff
@@ -0,0 +1,29 @@
...
+ * limitations under the License.
+ */
+package org.apache.beam.sdk.io;
+
+/**
+ * File system interface in Beam.
+ *
+ * <p>It defines APIs for writing file systems agnostic code.
+ *
+ * <p>All methods are protected, and they are for file system providers to implement.
```
Comment (lukecwik) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/io/FileSystem.java]: Any methods that you add to the interface are automatically public, with an abstract base class you can make the visibility protected and then this comment is not needed. It also protects users implementations if their visibility is right.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/io/FileSystems.java [hunk 24]
```diff
@@ -0,0 +1,143 @@
...
+/**
+ * Clients facing {@link FileSystem} utility.
+ */
+public class FileSystems {
+
+  public static final String DEFAULT_SCHEME = "default";
+
+  private static final Pattern URI_SCHEME_PATTERN = Pattern.compile("^[a-zA-Z][-a-zA-Z0-9+.]*$");
+
+  private static final Map<String, FileSystemRegistrar> SCHEME_TO_REGISTRAR = new ConcurrentHashMap();
```
Comment (lukecwik) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/io/FileSystems.java]: `ConcurrentHashMap() -> ConcurrentHashMap<>()`

File: sdks/java/core/src/main/java/org/apache/beam/sdk/io/FileSystems.java [hunk 25]
```diff
@@ -0,0 +1,143 @@
...
+
+  private static final Map<String, PipelineOptions> SCHEME_TO_DEFAULT_CONFIG = new ConcurrentHashMap();
```
Comment (lukecwik) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/io/FileSystems.java]: `ConcurrentHashMap() -> ConcurrentHashMap<>()`

File: sdks/java/io/hdfs/src/main/java/org/apache/beam/sdk/io/hdfs/HadoopFileSystemRegistrar.java (hunk 21 above)
Comment (lukecwik) [inline sdks/java/io/hdfs/src/main/java/org/apache/beam/sdk/io/hdfs/HadoopFileSystemRegistrar.java]: What is the meaning of "direct" FileSystem for you?

File: sdks/java/core/src/main/java/org/apache/beam/sdk/io/FileSystems.java (hunk 1 above)
Comment (lukecwik) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/io/FileSystems.java]: I'm referring to a class' static block, e.g.:
public class MyClass {
  static {
    // runs on class load
  }

Similar to the static block in PipelineOptionsFactory here:
https://github.com/apache/incubator-beam/blob/master/sdks/java/core/src/main/java/org/apache/beam/sdk/options/PipelineOptionsFactory.java#L483

File: sdks/java/core/src/main/java/org/apache/beam/sdk/io/FileSystems.java [hunk 26]
```diff
@@ -0,0 +1,143 @@
...
+    }
+  }
+
+  @VisibleForTesting
+  static void verifySchemesAreUnique(Set<FileSystemRegistrar> registrars) {
+    Multimap<String, FileSystemRegistrar> registrarsBySchemes =
+        TreeMultimap.create(Ordering.<String>natural(), Ordering.arbitrary());
+
+    for (FileSystemRegistrar registrar : registrars) {
+      registrarsBySchemes.put(registrar.getScheme(), registrar);
```
Comment (lukecwik) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/io/FileSystems.java]: Forgetting to lower case the scheme here before inserting it into the registrars map.

**Comment by lukecwik** [comment]
Just giving a heads up that I commented on some of your outdated comments so it will be worthwhile to look through them.

File: sdks/java/core/src/test/java/org/apache/beam/sdk/io/FileSystemsTest.java [hunk 27]
```diff
@@ -0,0 +1,76 @@
...
+import java.net.URI;
+
+import static org.junit.Assert.assertTrue;
+
+/**
+ * Tests for {@link FileSystems}.
+ */
+@RunWith(JUnit4.class)
+public class FileSystemsTest {
+
```
Comment (lukecwik) [inline sdks/java/core/src/test/java/org/apache/beam/sdk/io/FileSystemsTest.java]: Add tests for the various methods for upper/lower case schemes especially the duplicate scheme validation logic.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/io/FileSystem.java (hunk 23 above)
Comment (peihe) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/io/FileSystem.java]: changed to public abstract class (cannot make it protected, so still need to keep the comment).

**Comment by peihe** [comment]
Fixed tests.
PTAL

File: sdks/java/core/src/main/java/org/apache/beam/sdk/io/FileSystem.java (hunk 23 above)
Comment (lukecwik) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/io/FileSystem.java]: What are you trying to make protected?
What is blocking you from making the methods protected?

File: sdks/java/core/src/test/java/org/apache/beam/sdk/io/FileSystemsTest.java [hunk 28]
```diff
@@ -40,6 +41,14 @@
 
   @Test
   public void testSetDefaultConfig() throws Exception {
+    FileSystems.setDefaultConfig("file", PipelineOptionsFactory.create());
+    FileSystems.setDefaultConfig("FILE", PipelineOptionsFactory.create());
```
Comment (lukecwik) [inline sdks/java/core/src/test/java/org/apache/beam/sdk/io/FileSystemsTest.java]: This test seems to be missing validation

File: sdks/java/core/src/main/java/org/apache/beam/sdk/io/FileSystem.java (hunk 23 above)
Comment (peihe) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/io/FileSystem.java]: we need this abstract class be public, since it needs to be extends out side this package.
I will mark all methods to be protected, this comment is to tell clients that they should use FileSystems.

also, abstract class cannot be marked as protected:
http://stackoverflow.com/questions/8160019/why-cant-you-have-a-protected-abstract-class-in-java

**Comment by peihe** [comment]
PTAL
added validation for testSetDefaultConfig()

**Comment by lukecwik** [comment]
LGTM, will merge after test runs.
